File: billiards/app/services/tariffs.py
"""Операции с тарифами."""

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.models.journal import JournalEvent
from app.models.tariff import Tariff
from app.services.errors import ConflictError, NotFoundError
from app.services.journal import log_event


def list_tariffs(db: Session, *, only_active: bool = False) -> list[Tariff]:
    query = select(Tariff).order_by(Tariff.id)
    if only_active:
        query = query.where(Tariff.is_active.is_(True))
    return list(db.scalars(query))


def get_tariff(db: Session, tariff_id: int) -> Tariff:
    tariff = db.get(Tariff, tariff_id)
    if tariff is None:
        raise NotFoundError(f"Тариф id={tariff_id} не найден")
    return tariff


def create_tariff(db: Session, name: str, price_per_hour: int) -> Tariff:
    name = name.strip()
    if not name:
        raise ConflictError("Название тарифа не может быть пустым")
    if price_per_hour <= 0:
        raise ConflictError("Цена тарифа должна быть больше нуля")
    exists = db.scalar(select(Tariff).where(Tariff.name == name))
    if exists is not None:
        raise ConflictError(f"Тариф с названием «{name}» уже существует")
    tariff = Tariff(name=name, price_per_hour=price_per_hour)
    db.add(tariff)
    db.flush()
    log_event(
        db,
        JournalEvent.TARIFF_CREATED,
        f"Создан тариф «{tariff.name}» — {tariff.price_per_hour} ₽/час",
    )
    db.commit()
    return tariff

File: billiards-node/src/services/tariffs.js
// Операции с тарифами.

import { utcNow, withTransaction } from "../db.js";
import { ConflictError, NotFoundError } from "./errors.js";
import { JournalEvent, logEvent } from "./journal.js";

const toTariff = (row) => ({ ...row, is_active: Boolean(row.is_active) });

/**
 * @param {import("node:sqlite").DatabaseSync} db
 * @param {{onlyActive?: boolean}} [options]
 */
export function listTariffs(db, { onlyActive = false } = {}) {
  const rows = db
    .prepare(
      `SELECT id, name, price_per_hour, is_active FROM tariffs
       ${onlyActive ? "WHERE is_active = 1" : ""} ORDER BY id`
    )
    .all();
  return rows.map(toTariff);
}

/**
 * @param {import("node:sqlite").DatabaseSync} db
 * @param {number} tariffId
 */
export function getTariff(db, tariffId) {
  const row = db
    .prepare("SELECT id, name, price_per_hour, is_active FROM tariffs WHERE id = ?")
    .get(tariffId);
  if (!row) {
    throw new NotFoundError(`Тариф id=${tariffId} не найден`);
  }
  return toTariff(row);
}

/**
 * @param {import("node:sqlite").DatabaseSync} db
 * @param {string} name
 * @param {number} pricePerHour рублей в час, целое > 0
 */
export function createTariff(db, name, pricePerHour) {
  const trimmed = String(name ?? "").trim();
  if (!trimmed) {
    throw new ConflictError("Название тарифа не может быть пустым");
  }
  if (!Number.isInteger(pricePerHour) || pricePerHour <= 0) {
    throw new ConflictError("Цена тарифа должна быть целым числом больше нуля");
  }
  const exists = db.prepare("SELECT id FROM tariffs WHERE name = ?").get(trimmed);
  if (exists) {
    throw new ConflictError(`Тариф с названием «${trimmed}» уже существует`);
  }
  const id = withTransaction(db, () => {
    const { lastInsertRowid } = db
      .prepare(
        "INSERT INTO tariffs (name, price_per_hour, is_active, created_at) VALUES (?, ?, 1, ?)"
      )
      .run(trimmed, pricePerHour, utcNow());
    logEvent(
      db,
      JournalEvent.TARIFF_CREATED,
      `Создан тариф «${trimmed}» — ${pricePerHour} ₽/час`
    );
    return Number(lastInsertRowid);
  });
  return getTariff(db, id);
}
